### DISCUSSION TITLE
Building a DAO using RGB

### DISCUSSION THREAD

**Discussion by pxr64** [RGB-WG/rgb#55] [Q&A]
I want to get started on building a DAO using the blockchain network, would RGB be a suitable tool for getting something like that to production?

Also :

1. What are the challenges in storing data using the lightning network In terms of cost, performance and accessibility?
2. How should I  approach implementing  NFT’s or DAO system to lightning ( using RGB, TARO, or TAPSCRIPT)?
3. Would it be challenging to scale a system such as a DAO on the lightning network?

**Comment by dr-orlovsky** [comment]
Hi! I think we are still quite early: no toolchain suitable for external devs is there; it will take few more months to get it done. The protocol is there, but it is pretty low level.

**Reply by pxr64** [reply]
Thank you so much for the reply.

So we should start looking into building it from the ground up. 
I was thinking of approaching this in the following way:

1. Manage digital identity by distributing special key pairs to dao members.
2. To vote, the DAO members would have to sign special voting messages using the keys that were provided.

Now, I'm trying to figure out how to implement membership revocation and also prevent voting fraud and abuse. All this needs
to be decentralized and transparent, this is why I was looking into the option of embedding membership on the BTC chain somehow but I'm having a hard time finding a good solution.

Do you have any insight on this? It would be nice to have a smart contract-like piece of script which you can query to check if somebody is a member or not.

**Reply by dr-orlovsky** [reply]
I'd like to start with thinking from scratch about DAOs -- and in which parts they actually need to touch blockchain at all.

The heuristic here is that we need single-use-seals (and, thus, blockchain/RGB) only when we need to prove:
1) that the event has happened;
2) that the event happened once and only once.

For instance, key revocation in digital identity protocols requires (1) - and that's why we have [RGB22 interface](https://github.com/LNP-BP/LNPBPs/blob/master/lnpbp-0022.md).

So, for DAOs we need:
- digital identity with provable and detectable revocations - and we will have it with RGB22 identities;
- voting, where we need finality and provable uniquiness (that nobody has casted the vote twice, such that there could be multiple valid, but different outcomes of the voting ceremony);
- audit log record of the decision taken, which must be a chain of provably final events (like votings from the above);
- all other data may be kept offchain and committed into audit log record.

A primer for an audit log can be found in [RGB23](https://github.com/LNP-BP/LNPBPs/blob/master/lnpbp-0023.md). So basically all components are there, the last missing piece is an RGB contract interface which would cover main DAO workflows in a more organized fashion. I will work on a proposal in Feb at some point and post an update here.

**Reply by pxr64** [reply]
Nice, thank you for the insight.

What would be the preferred approach for storing the audit logs on-chain? 

I know we can embed information on the BTC base layer, but this can get quite expensive and it's really slow for the end user. On the other hand, if we store the information via lightning, wouldn't it get lost once the channel closes?

**Reply by dr-orlovsky** [reply]
> What would be the preferred approach for storing the audit logs on-chain?

not to store them on-chain :)
For which reason do you need them there?

It is in the interests of the DAO participants to keep the data on their client-side - and lightning is not "storing" the data (they are stored locally by RGB node) but just a mere data propagation layer. BTW you can use RGB RPC instead of lightning for data propagation, for instance.

**Reply by pxr64** [reply]
Having the client software store date is a good idea, it's a bit unclear how we make sure that the data is tamper-proof unless it's embedded on a chain in some way. 

I figured that we need to store some key information on the chain so that this info becomes the single source of truth. For example, when a member votes, they embed their intent in an on-chain transaction, later, after the voting ceremony ends, the outcome can be determined by looking at these transactions. 

However, maybe I'm thinking about this the wrong way because initially, I was envisioning the whole system as a special wallet application that would have the classic btc wallet features plus some features that would allow users to use interact with each other similar to how they would do if they were part of a DAO. 

These features essentially boil down to what you described earlier (that events are uniquely identifiable and immutable). This leads me to the idea that the wallet needs to be able to embed information on the chain and be able to access it later.

**Reply by dr-orlovsky** [reply]
I think you need to check how RGB and client-side validation is working - your question is about the exact problem they solve.

Please check https://www.youtube.com/watch?v=gGPLYfWOb_8

There is no point making DAO on RGB without using its advantages over on-chain storage.

**Reply by dr-orlovsky** [reply]
I think you need to check how RGB and client-side validation is working - your question is about the exact problem they solve.

Please check https://www.youtube.com/watch?v=gGPLYfWOb_8

There is no point making DAO on RGB without using its advantages over on-chain storage.

**Reply by cryptoquick** [reply]
@brinzi You might benefit from the solutions we're developing at diba.io to geographically distribute, periodically verify integrity, and checkpoint data hashes on the blockchain that we're developing to support RGB21 contracts, code, and content using [Carbonado](https://github.com/diba-io/carbonado). This allows keeping the RGB23 DAO log data in a safe, verifiable, distributed, and scalable place off-chain, which also expands what metadata and capabilities can be validated using RGB consensus checking.

**Reply by pxr64** [reply]
@cryptoquick  Amazing. Thanks for the reference, let me digest it a bit and I will leave some feedback. 
Also @dr-orlovsky, I really enjoyed your presentation, it gave me a lot of valuable insight, thank you. 🙏
